### DISCUSSION TITLE
C# equivalent of XAML compiled binding

### DISCUSSION THREAD

**Discussion by tranb3r** [CommunityToolkit/Maui#1447] [Q&A] [answered]
Compiled bindings improve data binding performance in .NET MAUI applications.
But all code examples for `Compiled bindings` are written in XAML.
Is `Compiled bindings` a XAML feature? What is the equivalent in C#?

**Comment by TheCodeTraveler** [comment, marked as answer]
For Compiled Bindings in C#, you need to install the [`CommunityToolkit.Maui.Markup` NuGet Package](https://www.nuget.org/packages/CommunityToolkit.Maui.Markup/).

### Docs 

https://learn.microsoft.com/dotnet/communitytoolkit/maui/markup/extensions/bindable-object-extensions#one-way-binding

### Example

```csharp
using static CommunityToolkit.Maui.Markup.GridRowsColumns;

class SampleContentPage : ContentPage
{
    public SampleContentPage()
    {
        Content = new Grid
        {
            RowDefinitions = Rows.Define(
                (Row.TextEntry, 36)),

            ColumnDefinitions = Columns.Define(
                (Column.Description, Star),
                (Column.Input, Stars(2))),

            Children =
            {
                new Label()
                    .Text("Code:")
                    .Row(Row.TextEntry).Column(Column.Description),

                new Entry
                {
                    Keyboard = Keyboard.Numeric,
                    BackgroundColor = Colors.AliceBlue,
                }.Row(Row.TextEntry).Column(Column.Input)
                 .FontSize(15)
                 .Placeholder("Enter number")
                 .TextColor(Colors.Black)
                 .Height(44)
                 .Margin(5, 5)
                 .Bind(Entry.TextProperty, static (ViewModel vm) vm => vm.RegistrationCode)
            }
        };
    }

    enum Row { TextEntry }
    enum Column { Description, Input }
}
```

**Comment by tranb3r** [comment]
Ok, this code is indeed "compiled", and not using a `string path`.
However, under the hood, is it 100% equivalent to the code generated by XAML compiled bindings?
How do they compare in terms of performance? I have the impression that XAML compiled bindings is a lot faster than this C# extension, but maybe I'm wrong.

**Reply by TheCodeTraveler** [reply]
> However, under the hood, is it 100% equivalent to the code generated by XAML compiled bindings

Yes, I wrote the code myself. Both use `TypedBindings` under the hood. It's all open source: https://github.com/CommunityToolkit/Maui.Markup/blob/main/src/CommunityToolkit.Maui.Markup/TypedBindingExtensions.cs

The performance is much better because it doesn't use reflection.

**Reply by tranb3r** [reply]
Actually, I don't see where it's using `TypedBinding`. Maybe you copied the code from dotnet maui and renamed it?
Also, are you saying it's much better than standard C# bindings (or standard XAML bindings), or is it much better than XAML compiled bindings?
I will probably need to run a benchmark to be convinced... ;)

**Comment by tranb3r** [comment]
Hi Brandon,

I've done more testing of compiled/typed bindings, both with xaml and c#.

Here are my conclusions: 
- as expected, when setting a view model property, the performance of bindings is the same between xaml compiled bindings and c# typed bindings, and it's better than when not using compiled/typed bindings.
- however, the performance of view creation is a lot worse when using c# typed bindings compared to any other option (xaml without compiled bindings, xaml with compiled bindings, c# without typed bindings).

Here is the code I've used to benchmark the different cases: https://github.com/tranb3r/Issues/tree/main/BenchmarkBindings

Here are the results of my benchmarks (higher is better):
![Screenshot 2023-10-09 173232](https://github.com/CommunityToolkit/Maui/assets/12516566/3b274094-51b3-4da6-a8fd-178c26927d18)

So, here are my questions:
- am I doing something wrong?
- why is the performance of c# typed bindings so bad on view creation?
- can we get the best of both worlds? (the performance of xaml compiled bindings, with the c# syntax)

**Reply by TheCodeTraveler** [reply]
Don’t use StopWatch to benchmark tests, instead use [`BenchmarkDotnet`](https://www.nuget.org/packages/BenchmarkDotNet).

The time required to instantiate the class is negligible compared to how long it takes for the View to receive an update following a `NotifyPropertyChanged` event on the view model. 

In a real-world .NET MAUI app, the ViewModel is instantiated once, while many many `NotifyPropertyChanged` events will fire for each ViewModel.

**Reply by TheCodeTraveler** [reply]
You’re more than welcome to dive into the source code for `CommunityToolkit.Maui.Markup` and submit a PR for the performance improvements you’re able to find!

**Reply by tranb3r** [reply]
For view creation, I get exactly the same results with benchmarkdotnet:
| Method                  | Mean     | Error    | StdDev   |
|------------------------ |---------:|---------:|---------:|
| RunXaml                 | 140.8 us |  2.66 us |  5.62 us |
| RunXamlCompiledBindings | 126.3 us |  2.52 us |  3.62 us |
| RunCSharp               | 139.2 us |  2.73 us |  4.09 us |
| RunCSharpTypedBindings  | 685.5 us | 13.07 us | 13.42 us |

**Reply by tranb3r** [reply]
So, in the end, I was wrongly assuming that everything you do in XAML, you can do in C#. 
This is clearly wrong: compiled Bindings is a XAML feature, and there is no C# equivalent in dotnet maui.

The Typed Bindings extensions give the same runtime performance for viewmodel updates, and offer compile-time types checking.
But view instantiation performance is impacted negatively. 
This is something I can see in my app. So it's wrong to assume that it's negligible in a real-world maui app.

I'm happy with XAML, and honestly I don't have the time to work on perf improvements of the framework. Actually I'm not even sure it's possible to improve this without modifying dotnet maui.

However, I think the documentations could be improved:
- for dotnet maui: it's not super clear that Compiled Bindings is a XAML feature with no C# equivalent in dotnet maui, meaning that if you want performance, you should stick with C#. 
- for communitytoolkit.maui: it's not super clear that Types Bindings extensions is not identical to XAML compiled bindings, and impacts negatively the performance of view instantiations.

**Reply by pictos** [reply]
@tranb3r can you share your benchmark? Looks like the repo you linked isn't using the BenchmarkDotnet

**Reply by tranb3r** [reply]
Hi Pedro,

I've updated my repro with BenchmarkDotNet in the Maui app.

For the CreateView test, the results are the same between StopWatch and BenchmarkDotNet.
This confirms the performance issue with C# Typed Bindings on View instantiation.

Results for CreateView (StopWatch):
| Method                  | Mean     |
|------------------------ |---------:|
| RunXaml                 | 280 us   |
| RunXamlCompiledBindings | 260 us   |
| RunCSharp               | 280 us   |
| RunCSharpTypedBindings  | 600 us   |

Results for CreateView (BenchmarkDotNet):
| Method                  | Mean     | Error    | StdDev   |
|------------------------ |---------:|---------:|---------:|
| RunXaml                 | 140.8 us |  2.66 us |  5.62 us |
| RunXamlCompiledBindings | 126.3 us |  2.52 us |  3.62 us |
| RunCSharp               | 139.2 us |  2.73 us |  4.09 us |
| RunCSharpTypedBindings  | 685.5 us | 13.07 us | 13.42 us |

For the UpdateViewModel test, the results are different between StopWatch and BenchmarkDotNet.
The StopWatch results seem coherent with the expected behavior of bindings (compiled or not, xaml or c#).
But the BenchmarkDotNet results seem wrong (no difference between the different tests).
So, I'm wondering if I'm measuring things correctly. Probably I'm not capturing the View update correctly.
What do you think? How could I improve this benchmark?

Results for UpdateViewModel (StopWatch):
| Method                  | Mean     |
|------------------------ |---------:|
| RunXaml                 | 20 us    |
| RunXamlCompiledBindings | 16 us    |
| RunCSharp               | 18 us    |
| RunCSharpTypedBindings  | 16 us    |

Results for UpdateViewModel (BenchmarkDotnet):
| Method                  | Mean     | Error    | StdDev   |
|------------------------ |---------:|---------:|---------:|
| RunXaml                 | 42.19 us | 0.839 us | 1.595 us |
| RunXamlCompiledBindings | 41.86 us | 0.836 us | 2.002 us |
| RunCSharp               | 41.55 us | 0.830 us | 1.049 us |
| RunCSharpTypedBindings  | 41.93 us | 0.831 us | 1.943 us |

**Reply by tranb3r** [reply]
@pictos 
Did you have the time to look at my benchmark?
Any suggestion on how to improve it?
Thanks

**Reply by pictos** [reply]
@tranb3r not yet, I'm afraid that I'll have time to take a good look on it after .net conf

**Reply by tranb3r** [reply]
@pictos I hope my follow-up isn't too bothersome. 
Your insights on this question would be highly valued whenever you have a moment. 
Thank you!

**Comment by CliffAgius** [comment]
Closed as answered

**Reply by tranb3r** [reply]
@CliffAgius 
Why are you closing this discussion?
I'm still needing help to figure out how to properly benchmark bindings.

**Reply by CliffAgius** [reply]
Sorry I am cleaning up the Discussions board as per the Community Stand-up discussion here: https://youtu.be/N9wMcBP4jtg?t=2889 and I see this as answered and it looked like the discussion was over.  I have re-opened again.

**Comment by CliffAgius** [comment]
Coming up to a year for this discussion is it ok to close?

**Reply by bijington** [reply]
I think @brminnick did something recently to improve the performance in C#

**Reply by TheCodeTraveler** [reply]
This added benchmarking, but I haven't taken the time to dig into performance improvements for `CommunityToolkit.Maui.Markup` yet: https://github.com/CommunityToolkit/Maui.Markup/pull/287

**Reply by TheCodeTraveler** [reply]
But yea, I think it’s fine to close this Discussion since the code in question is not in this repo. 

Folks can contribute over on https://github.com/CommunityToolkit/Maui.Markup/.

**Comment by CliffAgius** [comment]
Closed as per the discussion as answered.

**Reply by tranb3r** [reply]
@CliffAgius
Same as before: why are you closing this discussion again?
I'm still needing help to figure out how to properly benchmark bindings.

**Reply by CliffAgius** [reply]
Because @TheCodeTraveler above said the discussion was over and out was ok to close.
